test/snippet/argument_parser/validators_input_directory.cpp

```diff
@@ -0,0 +1,37 @@
+//! [usage]
+#include <seqan3/argument_parser/all.hpp>
+#include <seqan3/io/stream/debug_stream.hpp>
+#include <seqan3/std/filesystem>
+
+int main(int argc, const char ** argv)
+{
+    seqan3::argument_parser myparser{"Test", argc, argv}; // initialize
+
+    //! [validator_call]
+    std::filesystem::path mydir{};
+
+    myparser.add_option(mydir, 'd', "dir", "The directory containing the input files.",
+                        seqan3::option_spec::DEFAULT, seqan3::input_directory_validator{});
+    //! [validator_call]
+
+    // an exception will be thrown if the user specifies a directory that does not exists or has insufficient
+    // read permissions.
+    try
+    {
+        myparser.parse();
+    }
+    catch (seqan3::parser_invalid_argument const & ext) // the user did something wrong
+    {
+        std::cerr << "[PARSER ERROR] " << ext.what() << "\n"; // customize your error message
+        return -1;
+    }
+    catch (std::filesystem::filesystem_error const & ext)
+    {
+        std::cerr << "[IO ERROR] " << ext.what() << "\n"; // customize your error message
+        return -2;
+    }
```

Comment:
Hm, this is not really nice. I think all of these should be parser errors, it is not intuitive that one should catch IO-errors on argument parsing, right?

In general I would suggest:
  * make all internal interfaces just return values
  * have a single point (e.g. `operator()` of the validator) where you translate return values / error codes into exceptions.
  * always throw parser exceptions
  * only catch parser exceptions in the tutorials/snippets/...

If you are absolutely against this and want to have different exception types I would argue that we should change the tutorials/snippets/... to catch `std::exception` because having multiple blocks is ugly, difficult to explain and easy to get wrong.

Reply:
sorry, that I didn't make this clear earlier :see_no_evil: 
I didn't think about the implications of "what are the different exception types, let's deal with all of them".

Reply:
Ok, I agree that this is not beautiful. 

But I am also strongly against not forwarding errors from code that is not part of the validator checks. Like some low level IO errors because the filesystem was corrupted or whatever. 
To solve the issue that only one error type should be catched we  can throw_with_nested. In this case the upper exception is always a parser error. The user could then recursively print the nested exceptions if he wants to.

Reply:
I still think it would be nicer if all the interfaces return bool (or error_code) except the `operator()` and then you can throw parser exceptions with different texts depending on the function and the error_code (or you can even throw the error_code as part of the exception).

Catching and rethrowing is of course an option, too, it would at least guarantee that the top-level exception is always a parser exception. But it involves adding `catch` which we voted it against a while ago. But it's still better than the current solution.

I have a clear preference for the first, but I can live with the second. Let's not spend another week on this :wink:  But if you do do the second one, can you change the interfaces of the file resource to void functions that throw an error and not return some errors as `false` and some as exceptions? :pray: 

P.S: There is a proposal to do exceptions by values and we could actually implement that already by throwing error_codes and not types based on std::exceptions. We should discuss this a strategy meeting, but maybe after the release.

Reply:
>I still think it would be nicer if all the interfaces return bool (or error_code) except the operator() and then you can throw parser exceptions with different texts depending on the function and the error_code (or you can even throw the error_code as part of the exception).

Well, in general I am in favour of using both in combination (exception and error_codes). 
But this means to refactor the parser exceptions to define its own error_category and error codes. Then in case some other error happens its error code could be passed to the parser exception. The user can then easily examine the error code to see what went wrong.
But I am not refactoring this now.

> Catching and rethrowing is of course an option, too, it would at least guarantee that the top-level exception is always a parser exception. But it involves adding catch which we voted it against a while ago. But it's still better than the current solution.

In my opinion this would be the next best solution, which also keeps all information alive.

>I have a clear preference for the first, but I can live with the second. Let's not spend another week on this 😉 But if you do do the second one, can you change the interfaces of the file resource to void functions that throw an error and not return some errors as false and some as exceptions? 

They don't. They only throw on errors. Removing a non-existing file/directory is not an error. Hence they have the bool return value (or the max_int_t which I need to change).